feat: type extension

abemedia opened the pull request:
Add support for the extension `x-ogen-type`. E.g. `x-ogen-type: github.com/shopspring/decimal.Decimal` will result in the Go type of the schema being `decimal.Decimal`.

A few gotchas regarding the use:

- Types specified like this will always be (un)marshalled using the stdlib `json` package.
- If different external types have the same package name they will be aliased by adding integers e.g. `decimal1`, `decimal2` etc.
- Types from aliased packages will contain the name of the alias in the derived types e.g. `OptDecimal1Decimal`. For unique package names they will work the same as other types e.g. `OptDecimal`. Happy to remove this functionality though and resort to forcing users to use `x-ogen-name` when they have clashes if this feels too strange.

Closes #1335

tdakkota commented on gen/ir/constructors.go:
```diff
@@ -73,3 +79,22 @@ func Stream(name string, schema *jsonschema.Schema) *Type {
+func External(schema *jsonschema.Schema) (*Type, error) {
+	i := strings.LastIndex(schema.XOgenType, ".")
+	if i < 0 || i == len(schema.XOgenType)-1 {
+		return nil, fmt.Errorf("'%s' is not a valid Go type (expected 'pkg/path.Type')", schema.XOgenType)
```
We usually use go-faster/errors
```suggestion
		return nil, errors.Errorf("%q is not a valid Go type (expected 'pkg/path.Type')", schema.XOgenType)
```

abemedia replied:
Done.

tdakkota commented on gen/schema_gen.go:
```diff
@@ -266,6 +270,28 @@ func (g *schemaGen) generate2(name string, schema *jsonschema.Schema) (ret *ir.T
+		if err != nil {
+			return nil, errors.Wrap(err, "external type")
+		}
+
+		if pkg, ok := g.imports[path.Base(t.External.PackagePath)]; ok && pkg != t.External.PackagePath {
+			for i := 0; true; i++ {
+				importAlias := fmt.Sprintf("%s%d", path.Base(t.External.PackagePath), i)
+				if pkg, ok := g.imports[importAlias]; !ok || pkg == t.External.PackagePath {
+					t.External.ImportAlias = importAlias
+					t.Primitive = ir.PrimitiveType(importAlias + "." + t.External.GoName)
+					break
+				}
+			}
+		}
+
+		g.imports[cmp.Or(t.External.ImportAlias, path.Base(t.External.PackagePath))] = t.External.PackagePath
```
```suggestion
                pkgPath := t.External.PackagePath
                pkgName := path.Base(pkgPath)
                
		if pkg, ok := g.imports[pkgName]; ok && pkg != pkgPath {
			for i := 0; true; i++ {
				importAlias := fmt.Sprintf("%s%d", pkgName, i)
				if pkg, ok := g.imports[importAlias]; !ok || pkg == pkgPath {
					t.External.ImportAlias = importAlias
					t.Primitive = ir.PrimitiveType(importAlias + "." + t.External.GoName)
					break
				}
			}
		}

		g.imports[cmp.Or(t.External.ImportAlias, pkgName)] = pkgPath
```

abemedia replied:
Have amended although it's slightly different now as I added support for builtin types.

tdakkota commented on _testdata/positive/type_extension.yml:
```diff
@@ -0,0 +1,70 @@
+                    x-ogen-type: github.com/ogen-go/ogen/_testdata/types/foo/decimal.Decimal
+                  bar:
+                    type: string
+                    x-ogen-type: github.com/ogen-go/ogen/_testdata/types/bar/decimal.Decimal
```
It would be nice to have tests for arrays and maps (via `additionalProperties`) too

abemedia replied:
Done

tdakkota commented on internal/integration/generate.go:
```diff
@@ -34,6 +34,7 @@ package integration
 //
 //go:generate go run ../../cmd/ogen -v --clean -target test_enum_naming       ../../_testdata/positive/enum_naming.yml
 //go:generate go run ../../cmd/ogen -v --clean -target test_naming_extensions ../../_testdata/positive/naming_extensions.json
+//go:generate go run ../../cmd/ogen -v --clean -target test_type_extension ../../_testdata/positive/type_extension.yml
```
It would be nice to have a JSON decode-encode-decode test too.

abemedia replied:
Have added that for both JSON and params.

abemedia replied:
Actually the tests for the params only cover decoding, not encoding. Shall I write E2E tests to cover everything?

tdakkota reviewed (commented):
LGTM in general, but encoding is tricky, so we need a bit more tests.

> Types from aliased packages will contain the name of the alias in the derived types e.g. `OptDecimal1Decimal`. For unique package names they will work the same as other types e.g. `OptDecimal`. Happy to remove this functionality though and resort to forcing users to use `x-ogen-name` when they have clashes if this feels too strange.

Let's remove it, then.

tdakkota commented on gen/_template/header.tmpl:
```diff
@@ -58,6 +58,9 @@ import (
 	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
 	"go.opentelemetry.io/otel/trace"
 	"go.uber.org/multierr"
+	{{ range $alias, $pkg := $.Imports -}}
+	{{ $alias }} "{{ $pkg }}"
+	{{ end }}
```
No,[ `text/template` sorts keys before iteration](https://github.com/golang/go/blob/go1.23.0/src/text/template/exec.go#L410-L413). 

https://go.dev/play/p/KADXY95uWkC

abemedia commented on _testdata/positive/time_extension.yml:
This is slightly unrelated but I realised aliases were broken when using `x-ogen-time-format` so I updated the tests and fixed it.

abemedia commented on conv/decode.go:
```diff
@@ -71,8 +77,13 @@ func ToString(s string) (string, error) {
+
 func ToBytes(s string) ([]byte, error) {
-	return []byte(s), nil
+	return toBytes(s), nil
```
I took the liberty of changing this so it doesn't allocate memory.

abemedia replied:
But this is only used to convert URL parameters, isn't it? Those strings are only referenced briefly and then not used again as far as I can see.

tdakkota replied:
Those strings reference values of the map returned by `URL.Query()`. In turn, the keys and values returned by `Query()` would likely reference the raw URL inside `url.URL`. This raw URL may then be referenced by any middleware or by net/http itself.

I think it's quite possible it could corrupt something unexpectedly.

abemedia commented on conv/encode.go:
```diff
@@ -28,7 +34,7 @@ func Float64ToString(v float64) string { return strconv.FormatFloat(v, 'f', 10,
 
 func StringToString(v string) string { return v }
-func BytesToString(v []byte) string  { return string(v) }
+func BytesToString(v []byte) string  { return unsafe.String(unsafe.SliceData(v), len(v)) }
```
Again, unrelated but took the liberty to reduce allocs.

abemedia commented on gen/ir/external.go:
```diff
@@ -0,0 +1,131 @@
+	ExternalJSON
+	// ExternalText indicates that the type implements stdlib encoding.TextMarshaler or encoding.TextUnmarshaler.
+	ExternalText
+	// ExternalBinary indicates that the type implements stdlib encoding.BinaryMarshaler or encoding.BinaryUnmarshaler.
+	ExternalBinary
```
This currently isn't used. I was however considering adding support for `encoding.BinaryMarshaler` and `encoding. BinaryUnmarshaler` for values that have `type: string` and `format: byte`. What do you think?

abemedia commented on json/external.go:
```diff
@@ -0,0 +1,156 @@
+	}
+)
+
+// EncodeNative encodes a value using [Marshaler] interface.
+func EncodeNative[T any, P marshaler[T]](e *jx.Encoder, v T) {
+	P(&v).Encode(e)
+}
+
+// DecodeNative decodes a value using [Unmarshaler] interface.
+func DecodeNative[T any, P unmarshaler[T]](d *jx.Decoder) (T, error) {
+	var v T
+	return v, P(&v).Decode(d)
+}
+
+// EncodeStringNative encodes a string value using [Marshaler] interface.
+func EncodeStringNative[T any, P marshaler[T]](e *jx.Encoder, v T) {
+	EncodeNative[T, P](e, v)
+}
+
+// DecodeStringNative decodes a string value using [Unmarshaler] interface.
+func DecodeStringNative[T any, P unmarshaler[T]](d *jx.Decoder) (T, error) {
+	return DecodeNative[T, P](d)
+}
```
Like with `conv` there's a lot of duplication to handle (un)quoting strings.

abemedia commented on examples/ex_openapi/output.gen.go:
```diff
@@ -5890,16 +5890,16 @@ func (s *XMLPattern0) init() XMLPattern0 {
-		s.ExclusiveMaximum.SetTo(val)
+		val := bool(false)
+		s.ExclusiveMaximum.SetTo(ExclusiveMaximum(val))
 	}
 	{
-		val := ExclusiveMinimum(false)
-		s.ExclusiveMinimum.SetTo(val)
+		val := bool(false)
+		s.ExclusiveMinimum.SetTo(ExclusiveMinimum(val))
 	}
 	{
-		val := UniqueItems(false)
-		s.UniqueItems.SetTo(val)
+		val := bool(false)
+		s.UniqueItems.SetTo(UniqueItems(val))
 	}
 }
```
This is an unintended side-effect of the fix for aliases that have formats. Although it's not as pretty it doesn't have any effect on performance and keeps the template more readable. If however you feel strongly about this I can add some extra handling to the template to avoid it.

abemedia commented:
@tdakkota I made a bunch of changes and improved the (un)marshlaling of external values to avoid the reflection cost in the stdlib `json` package where possible. Could I get another review?

abemedia commented:
Actually hold off I just found a bug. Will fix and report back when done.

abemedia commented on gen/_template/header.tmpl:
```diff
@@ -26,47 +26,8 @@ import (
-	"encoding/base64"
-	"fmt"
-	"io"
-	"math"
-	"math/big"
-	"math/bits"
-	"mime"
-	"mime/multipart"
-	"net"
-	"net/http"
-	"net/netip"
-	"net/url"
-	"regexp"
-	"sort"
-	"strconv"
-	"strings"
-	"sync"
-	"time"
-
-	"github.com/go-faster/errors"
-	"github.com/go-faster/jx"
-	"github.com/google/uuid"
-	"go.opentelemetry.io/otel"
-	"go.opentelemetry.io/otel/attribute"
-	"go.opentelemetry.io/otel/codes"
-	"go.opentelemetry.io/otel/metric"
-	"go.opentelemetry.io/otel/metric"
-	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
-	"go.opentelemetry.io/otel/trace"
-	"go.uber.org/multierr"
-
-	"github.com/ogen-go/ogen/conv"
-	ht "github.com/ogen-go/ogen/http"
-	"github.com/ogen-go/ogen/middleware"
-	"github.com/ogen-go/ogen/json"
-	"github.com/ogen-go/ogen/ogenregex"
-	"github.com/ogen-go/ogen/ogenerrors"
-	"github.com/ogen-go/ogen/otelogen"
-	"github.com/ogen-go/ogen/uri"
-	"github.com/ogen-go/ogen/validate"
```
All imports are handled like the ones from the external packages to avoid clashes.

abemedia commented on gen/write.go:
```diff
@@ -21,6 +21,10 @@ import (
 	"github.com/ogen-go/ogen/ogenregex"
 )
 
+func init() {
+	imports.LocalPrefix = "github.com/ogen-go/ogen"
+}
```
I don't like this but it is to maintain the current grouping of imports in generated files. If you're not that bothered about that I'd be in favour of removing it as I'm not a fan of side-effects such as setting global vars is 3rd party packages.

tdakkota replied:
Let's remove it. Import grouping is not that important, I think.

abemedia replied:
Have removed it and regenerated everything.

abemedia commented:
@tdakkota I've resolved the bugs mentioned in the previous message and this is again ready for review.

tdakkota commented on _testdata/positive/time_extension.yml:
```diff
@@ -100,3 +114,11 @@ paths:
                     x-ogen-time-format: 2006-01-02T15:04:05.999999999Z07:00
+                  alias:
+                    $ref: '#/components/schemas/Alias'
+components:
+  schemas:
+    Alias:
+      type: string
+      format: date-time
+      x-ogen-time-format: 02/01/2006 3:04:05PM
```
Let's keep it here, I don't think it needs a separate test.

tdakkota commented on conv/decode.go:
```diff
@@ -203,6 +214,61 @@ func ToStringFloat64(s string) (float64, error) {
+func ToNative[T any, P ogenUnmarshaler[T]](s string) (T, error) {
+	var v T
+	err := P(&v).Decode(jx.DecodeBytes(toBytes(s)))
+	return v, err
+}
+
+func ToStringNative[T any, P ogenUnmarshaler[T]](s string) (T, error) {
+	return ToNative[T, P](strconv.Quote(s))
+}
+
+func ToText[T any, P textUnmarshaler[T]](s string) (T, error) {
+	var v T
+	err := P(&v).UnmarshalText(toBytes(s))
+	return v, err
+}
+
+func ToStringText[T any, P textUnmarshaler[T]](s string) (T, error) {
+	return ToText[T, P](s)
+}
+
+func ToJSON[T any, P jsonUnmarshaler[T]](s string) (T, error) {
+	var v T
+	err := P(&v).UnmarshalJSON(toBytes(s))
+	return v, err
+}
+
+func ToStringJSON[T any, P jsonUnmarshaler[T]](s string) (T, error) {
+	return ToJSON[T, P](strconv.Quote(s))
+}
+
+func ToExternal[T any](s string) (T, error) {
+	var v T
+	err := stdjson.Unmarshal(toBytes(s), &v)
+	return v, err
+}
+
+func ToStringExternal[T any](s string) (T, error) {
+	return ToExternal[T](strconv.Quote(s))
+}
+
```
> If you're not keen on these I can also handle (un)quoting in the templates and just have a single function for each type.

If it would not complicate the template significantly, it would be nice.

abemedia replied:
It wouldn't complicate it significantly. I could add a template helper function like `QuoteStrings() bool` to check if something needs (un)quoting.

abemedia replied:
I had a go at this and although it's not hard for the `conv` package, it isn't as straightforward for the `json` package, as we need to use different methods from `jx`. For strings we need to use `(*jx.Decoder).ByteStr` while for others we use `(*jx.Decoder).Raw` so I cannot think of a clean way to do this universally.
What I could do is avoid the pointless `json` functions that are just aliases by having the `Decoder` and `Encoder` functions only add the `String` prefix where it's really needed.

abemedia replied:
@tdakkota what do you reckon?

abemedia replied:
I have improved the template helpers to remove the pointless `StringXXX` functions.

abemedia commented on gen/ir/external.go:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	getIface := func(pkgs []*packages.Package, pkgPath, typeName string) *types.Interface {
+		return getType(pkgs, pkgPath, typeName).Underlying().(*types.Interface).Complete()
```
This function is only used for the interfaces we're checking in the maps above, not the types the user supplies.

abemedia commented:
@tdakkota sorry for the delay but I have just added another commit to improve the encode/decode by removing the pointless string functions and added support for binary types.
Mind giving it another review? If you're happy I'll squash the commits, just thought this makes it easier to review the actual changes.

abemedia commented on gen/ir/template_helpers.go:
```diff
@@ -79,7 +79,32 @@ func (t Type) uriFormat() string {
 }
 
+// externalType defines what available interface to use for encoding/decoding.
+func (t Type) externalType(e ExternalEncoding) ExternalEncoding {
+	switch {
+	case e.Has(ExternalNative):
+		return ExternalNative
+	case e.Has(ExternalJSON):
+		return ExternalJSON
+	case e.Has(ExternalText):
+		return ExternalText
+	case e.Has(ExternalBinary) && t.Schema.Type == jsonschema.String &&
+		(t.Schema.Format == "byte" || t.Schema.Format == "base64"):
+		return ExternalBinary
+	default:
+		return 0
+	}
+}
```
This function decides which of the available interfaces to use. Not sure if we should amend the order based on the caller though i.e. should we prefer `(Un)MarshalText` for URL params?

abemedia replied:
Done. See following comment for reasoning behind new order: https://github.com/ogen-go/ogen/pull/1460#pullrequestreview-2993546473

tdakkota reviewed (approved):
Well, it was a long road.

Huge thanks for all your contributions to the ogen.

abemedia commented:
Thank you!
